src/have.h

```diff
@@ -61,6 +61,12 @@
 #endif
 #endif
 
+#if defined(HAVE_ATTR_MALLOC_WITH_ARG)
+#define ATTR_CONSTRUCTOR(dealloc) __attribute__((__malloc__(dealloc)))
```

Comment:
The naming is a bit unfortunate because there is also an `__attribute__((constructor))`. I just checked and you can use empty parentheses even with compilers that don't support `__malloc__`-with-arg, so you can make a single macro that covers both cases:
```
#if defined(HAVE_ATTR_MALLOC_WITH_ARG)
#define ATTR_MALLOC(dealloc) __attribute__((__malloc__(dealloc)))
#else
#define ATTR_MALLOC(dealloc) __attribute__((__malloc__))
#endif
```
And then just use empty parentheses when using it on functions that don't have a corresponding deallocator, like so:
```
static inline void *xmalloc(size_t n) ATTR_MALLOC();
```

Reply:
I don't think we can do this since they're in effect two completely different attributes (which probably should have been reflected in their names, but I'm not going to argue with ~~gods~~ compiler developers).

afaik `malloc(deallocator)` means "user must call `deallocator` to release resources *returned* by this function" (not *created*, and not necessarily memory); so it's basically a marker to tie pairs of "allocate/deallocate". The PR was posted mostly for this attribute, since there were lots of mismatched calls previously (most were created by myself); let the compiler prevent it in the future.

`malloc` means "this function creates a new island of memory which doesn't have pointers into anything previously existing, thus resembling the behavior of `malloc()`"; it's just an optimization.

The first one can be useful in situations like `get_from_cache()`/`return_to_cache()`, where the second one doesn't apply.

I removed plain `malloc` from functions that return memory allocated by libraries like OpenSSL. Who knows what those opaque structures could contain, or where they could point to. (Also from functions that take a pointer argument, allocate a new struct, and embed that pointer into it, for obvious reasons).

---

https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html

> Attribute malloc indicates that a function is malloc-like, i.e., that the pointer P returned by the function cannot alias any other pointer valid when the function returns, and moreover no pointers to valid objects occur in any storage addressed by P.

> Independently, the form of the attribute with one or two arguments associates deallocator as a suitable deallocation function for pointers returned from the malloc-like function.

> Using the attribute with no arguments is designed to improve optimization by relying on the aliasing property it implies. Functions like malloc and calloc have this property because they return a pointer to uninitialized or zeroed-out, newly obtained storage. However, functions like realloc do not have this property, as they may return pointers to storage containing pointers to existing objects.

> Associating a function with a deallocator helps detect calls to mismatched allocation and deallocation functions ...

> To indicate that an allocation function both satisifies the nonaliasing property and has a deallocator associated with it, both the plain form of the attribute and the one with the deallocator argument must be used.

Reply:
Ugh, why did they give it the same name... Anyway, good explanation and I agree with the removal of `malloc` from things that are not guaranteed to have malloced exactly the pointer they return.